### ISSUE TITLE
How to run the Pyramid with albert_bert_embedding?

### ISSUE THREAD

**Issue by Nicozwy** [LorrinWWW/Pyramid#2]
Thanks for sharing the source code.
If I want to run the Pyramid with albert_bert_embedding, what should I do? Especially, how to determine the parameter of "--lm_emb_dim"?

**Comment by LorrinWWW** [comment]
You need to run "./run/gen_albert_bert_emb.py" first to generate the contextualized embeddings, and then run "./train_ner.py" to train the model.
For the second question, lm_emb_dim is the sum of hidden size of Albert and Bert. For example, if we use "albert-xxlarge-v2" and "bert-large-uncased", and given the hidden size of albert-xxlarge is 4096, and the hidden size of bert-large is 1024, "lm_emb_dim" here would be 4096+1024=5120.

**Comment by Nicozwy** [comment]
Thanks for your rapid responds! I have make it successly .

By the way, how to calculate the the possible number of IOB2 tags, and why the value of `tag_vocab_size` should be bigger than the possible number of IOB2 tags. 
For example, the value of `tag_vocab_size` is set to 60 by default when evaluating the model on ACE05 dataset, however, there are only 8 tags in this dataset totally, i.e., ["O", "FAC", "ORG", "LOC", "WEA", "GPE", "VEH", "PER"].

In addition, when using the adam optimizer, the model seems to have a poor performance than using sgd optimizer. why?

Could you explain those detailly.

**Comment by Nicozwy** [comment]
@LorrinWWW 
Thanks for your rapid responds! I have make it successly .

By the way, how to calculate the the possible number of IOB2 tags, and why the value of tag_vocab_size should be bigger than the possible number of IOB2 tags.
For example, the value of tag_vocab_size is set to 60 by default when evaluating the model on ACE05 dataset, however, there are only 8 tags in this dataset totally, i.e., ["O", "FAC", "ORG", "LOC", "WEA", "GPE", "VEH", "PER"].

In addition, when using the adam optimizer, the model seems to have a poor performance than using sgd optimizer. why?

Could you explain those detailly.

**Comment by Nicozwy** [comment]
@LorrinWWW 
Thanks for your rapid responds! I have made it successfully .

By the way, how to calculate the the possible number of IOB2 tags, and why the value of tag_vocab_size should be bigger than the possible number of IOB2 tags.
For example, the value of tag_vocab_size is set to 60 by default when evaluating the model on ACE05 dataset, however, there are only 8 tags in this dataset totally, i.e., ["O", "FAC", "ORG", "LOC", "WEA", "GPE", "VEH", "PER"].

In addition, when using the adam optimizer, the model seems to have a poor performance than using sgd optimizer. why?

Could you explain those detailly.

**Comment by LorrinWWW** [comment]
The value of tag_vocab_size should be either bigger than or equal to the number of IOB2 tags, which seems trivial to the performance. So setting a big value is easier, right? In addition, there are ["O", "B-FAC", "B-ORG", "B-LOC", "B-WEA", "B-GPE", "B-VEH", "B-PER", "I-FAC", "I-ORG", "I-LOC", "I-WEA", "I-GPE", "I-VEH", "I-PER"], although I-tags are relatively rare here.
It is common to see that SGD optimizer performs better than Adam (in terms of accuracy), but overall, the difference should be smaller than 1%. Have you tried to tune the learning rate for Adam? The learning rate for SGD is usually too big for Adam. You can try 1e-3, which may lead to a better performance.

**Comment by Nicozwy** [comment]
Thank you for your response.@LorrinWWW 
Could you explain the reason why Albert and bert were adopted to generate embeddings  simultaneously？

**Comment by LorrinWWW** [comment]
Actually, there is no particular reason for that. In Strakova et al. 2019, they found BERT+Flair performed better than BERT. Therefore, we tried other combinations, and found that ALBERT+BERT showed some improvements. We also tested ALBERT+BERT+Flair, but it had no further improvement.

**Comment by Nicozwy** [comment]
Thank you! 
For some reasons, we can not get access to the flair embedding models when run "./run/gen_bert_flair_emb.py". 

Could you share the files, i.e.,  `embeddings-v0.4.1/big-news-forward--h2048-l1-d0.05-lr30-0.25-20/news-forward-0.4.1.pt` and `embeddings-v0.4.1/big-news-backward--h2048-l1-d0.05-lr30-0.25-20/news-backward-0.4.1.pt`?@LorrinWWW

**Comment by LorrinWWW** [comment]
Check if [this](https://drive.google.com/drive/folders/1ZMCSPaAFUNaOZQrOwLcXL-xCtkqBED1s?usp=sharing) works. You may need to put them in "~/.flair/embeddings/"

**Comment by Nicozwy** [comment]
Thank you for your help.
